tokenizers/src/models/unigram/trainer.rs

```diff
@@ -94,39 +96,40 @@ impl UnigramTrainer {
         let mut min_score_penalty = 0.0;
         let min_score_penalty_delta = 0.0001;
 
-        let mut pieces: HashMap<String, f64> = HashMap::new();
+        let mut pieces: Vec<(String, f64)> = vec![];
+        let mut inserted: HashSet<String> = HashSet::new();
         let existing_pieces: HashMap<String, f64> = model.iter().cloned().collect();
-        // XXX: Make sure bos, eos and unk exists and are ids 0, 1, 2
-        pieces.insert(self.unk_token.clone(), 0.0);
         for c in required_chars {
             if let Some(t) = existing_pieces.get(&c) {
-                pieces.insert(c, *t);
+                inserted.insert(c.clone());
```

Comment:
Is the hashSet `inserted` really necessary ? `required_chars` is already a HashMap, so we shouldn't get duplicates anyway, no ?

Reply:
`inserted` has been added here because we are changing `pieces` to be a `Vec<(String, f64)>` instead of a `HashMap<String, f64>`, so we still need a way to check if a token has been inserted with `O(1)`

Reply:
My bad, did not see that change to `Vec`